pkg/controller/priorityqueue/priorityqueue_test.go

```diff
@@ -0,0 +1,486 @@
+		Expect(metrics.adds["test"]).To(Equal(1))
+		Expect(metrics.retries["test"]).To(Equal(0))
+	})
+
+	It("returns items in order", func() {
+		q, metrics := newQueue()
+		defer q.ShutDown()
+
+		q.AddWithOpts(AddOpts{}, "foo")
+		q.AddWithOpts(AddOpts{}, "bar")
+
+		item, _, _ := q.GetWithPriority()
+		Expect(item).To(Equal("foo"))
+		item, _, _ = q.GetWithPriority()
+		Expect(item).To(Equal("bar"))
+
+		Expect(metrics.depth["test"]).To(Equal(0))
+		Expect(metrics.adds["test"]).To(Equal(2))
+	})
+
+	It("doesn't return an item that is currently locked", func() {
+		q, metrics := newQueue()
+		defer q.ShutDown()
+
+		q.AddWithOpts(AddOpts{}, "foo")
+
+		item, _, _ := q.GetWithPriority()
+		Expect(item).To(Equal("foo"))
+
+		q.AddWithOpts(AddOpts{}, "foo")
+		q.AddWithOpts(AddOpts{}, "bar")
+		item, _, _ = q.GetWithPriority()
+		Expect(item).To(Equal("bar"))
+
+		Expect(metrics.depth["test"]).To(Equal(1))
+		Expect(metrics.adds["test"]).To(Equal(3))
+	})
+
+	It("returns an item as soon as its unlocked", func() {
+		q, metrics := newQueue()
+		defer q.ShutDown()
+
+		q.AddWithOpts(AddOpts{}, "foo")
+
+		item, _, _ := q.GetWithPriority()
+		Expect(item).To(Equal("foo"))
+
+		q.AddWithOpts(AddOpts{}, "foo")
+		q.AddWithOpts(AddOpts{}, "bar")
+		item, _, _ = q.GetWithPriority()
+		Expect(item).To(Equal("bar"))
+
+		q.AddWithOpts(AddOpts{}, "baz")
+		q.Done("foo")
+		item, _, _ = q.GetWithPriority()
+		Expect(item).To(Equal("foo"))
+
+		Expect(metrics.depth["test"]).To(Equal(1))
+		Expect(metrics.adds["test"]).To(Equal(4))
+	})
+
+	It("de-duplicates items", func() {
+		q, metrics := newQueue()
+		defer q.ShutDown()
+
+		q.AddWithOpts(AddOpts{}, "foo")
+		q.AddWithOpts(AddOpts{}, "foo")
+
+		Consistently(q.Len).Should(Equal(1))
+
+		cwq := q.(*priorityqueue[string])
+		cwq.lockedLock.Lock()
+		Expect(cwq.locked.Len()).To(Equal(0))
+
+		Expect(metrics.depth["test"]).To(Equal(1))
+		Expect(metrics.adds["test"]).To(Equal(1))
+	})
+
+	It("retains the highest priority", func() {
+		q, metrics := newQueue()
+		defer q.ShutDown()
+
+		q.AddWithOpts(AddOpts{Priority: 1}, "foo")
+		q.AddWithOpts(AddOpts{Priority: 2}, "foo")
+
+		item, priority, _ := q.GetWithPriority()
+		Expect(item).To(Equal("foo"))
+		Expect(priority).To(Equal(2))
+
+		Expect(q.Len()).To(Equal(0))
+
+		Expect(metrics.depth["test"]).To(Equal(0))
+		Expect(metrics.adds["test"]).To(Equal(1))
+	})
+
+	It("gets pushed to the front if the priority increases", func() {
+		q, metrics := newQueue()
+		defer q.ShutDown()
+
+		q.AddWithOpts(AddOpts{}, "foo")
+		q.AddWithOpts(AddOpts{}, "bar")
+		q.AddWithOpts(AddOpts{}, "baz")
+		q.AddWithOpts(AddOpts{Priority: 1}, "baz")
+
+		item, priority, _ := q.GetWithPriority()
+		Expect(item).To(Equal("baz"))
+		Expect(priority).To(Equal(1))
+
+		Expect(q.Len()).To(Equal(2))
+
+		Expect(metrics.depth["test"]).To(Equal(2))
+		Expect(metrics.adds["test"]).To(Equal(3))
+	})
+
+	It("retains the lowest after duration", func() {
```

Comment:
Not sure how this test verifies that

Reply:
By checking that it immediately returns the object and the queue is empty afterwards?

Reply:
Nevermind. Not sure what I was thinking :)